#164957460 Setup email configuration

NedyUdombat opened the pull request:
#### What does this PR do?
Setup email configuration

#### Description of Task to be completed?
* Install nodemailer
* create a configuration file for email that uses nodemailer and smtp
* add smtp details example to the enviroment example file

#### How should this be manually tested?
No test measure available yet

#### Any background context you want to provide?
* Create a mailtrap account at mailtrap.io 
* copy the changes from the .env.example into your env.
* copy the credentials from the smtp settings into your `.env`

To use the configuration
import `sendMail` from `mailConfig.js`
call the function and pass in the following parameters where available
```
const mailBody = {
  receiver:[],
  subject: '', 
  text: '', 
  html: '',
};

sendMail(mailBody);
```

#### What are the relevant pivotal tracker stories? (if applicable)
#164957460

#### Screenshots
None

NedyUdombat committed:
[chore #164957460] Setup email configuration

Install nodemailer
create a configuration file for email that uses nodemailer and smtp
add smtp details example to enviroment example file
add IDE folder to gitignore

[start #164957469]

NedyUdombat committed:
[chore #164957460] Remove console.log statement

Japhethca commented on src/config/mailConfig.js:
```diff
@@ -0,0 +1,39 @@
+});
+
+const sendMail = (req, res, body) => {
+  const allRecipientEmail = body.receiver.join(', ');
```
De-structure params from body object. 
`const { email } = res.body;`

NedyUdombat replied:
This has been resolved

Japhethca commented on src/config/mailConfig.js:
```diff
@@ -0,0 +1,39 @@
+});
+
+const sendMail = (req, res, body) => {
+  const allRecipientEmail = body.receiver.join(', ');
```
use `receivers` instead of `receiver`

NedyUdombat committed:
[chore #164957460]Implement PR feedback

Destructure body object
rename receiver to receivers

[start #164957460]

Japhethca commented on src/config/mailConfig.js:
```diff
@@ -0,0 +1,42 @@
+    html,
+  };
+
+  transporter.sendMail(mailOptions, (error, info) => {
```
let's use `async/await` here instead of callbacks. 
```javacript
try {
  let info = await transporter.sendMail(mailOptions);
} catch (error) {
  return res.json({error});
}
return res.json({info});
```

NedyUdombat replied:
noted

Japhethca replied:
can you also write tests for this while you at that?

NedyUdombat replied:
okay

Japhethca reviewed (changes requested):
Just one more comment.

NedyUdombat committed:
[chore #164957460]Implement PR feedback

add .DS_STORE file to gitignore
install dependecies for mock testing of email
write test for email
convert email fucntion to use async await

[start #164957460]

Japhethca commented on src/tests/config/mailConfig.test.js:
```diff
@@ -0,0 +1,47 @@
+      expect(info.response).to.eql('250 2.0.0 Ok: queued');
+      expect(info.envelope).to.be.a('object');
+      expect(info.messageId).to.be.a('string');
+    }
```
add test for when there is an error sending emails.

NedyUdombat replied:
test has been added

Japhethca commented on src/tests/config/mailConfig.test.js:
```diff
@@ -0,0 +1,47 @@
+  receivers,
+  subject: 'Test mail',
+  text: '',
+  html: '<p>Test mail to verify if email configuration is working</p>'
```
add comma here.

NedyUdombat replied:
the comma has been added

Japhethca commented on src/config/mailConfig.js:
```diff
@@ -0,0 +1,40 @@
+  port,
+  auth: {
+    user: username,
+    pass
```
add comma here.

NedyUdombat replied:
added

NedyUdombat committed:
[chore #164957460]Implement PR feedback

Implement more test cases
add commas to the last element of an object
throw error caught by the catch block
remove mailOption decxlaration from the try block

[start #164957460]

NedyUdombat commented on src/tests/config/mailConfig.test.js:
```diff
@@ -0,0 +1,69 @@
+
+  it('should fail to send an email if there is no object passed to the sendMail function', async () => {
+    try {
+      info = await sendMailer(noBody);
```
`noBody` is not defined for a test case to test for when an undefined variable is passed into the `sendMail()` function.
